task/source-build/0.1/source-build.yaml

```diff
@@ -0,0 +1,472 @@
+                symlink_member = ""
+                with tarfile.open(f"{base_sources_extraction_dir}/{digest}", "r:gz") as tar:
+                    for member in tar:
+                        if member.isfile():
+                            blob_member = member.name
+                        elif member.issym():
+                            symlink_member = member.name
+
+                if symlink_member.startswith("./rpm_dir/"):
+                    dest_dir = sib_dirs.rpm_dir
+                    log.debug("Prepare SRPM %s", symlink_member)
+
+                    extract_blob_member(digest, blob_member, dest_dir,
+                                        rename_to=os.path.basename(symlink_member),
+                                        work_dir=base_sources_extraction_dir,
+                                        log=log)
+
+                elif symlink_member.startswith("./extra_src_dir/"):
+                    extra_src_archive = os.path.basename(symlink_member)
+                    log.debug("Prepare extra source %s", extra_src_archive)
+                    # one extra source archive per directory, no matter what the directory name is.
+                    dest_dir = create_dir(extra_src_dir, extra_src_archive)
+                    sib_dirs.extra_src_dirs.append(dest_dir)  # save this directory for bsi
+
+                    extract_blob_member(digest, blob_member, str(dest_dir),
+                                        rename_to=extra_src_archive,
+                                        work_dir=base_sources_extraction_dir,
+                                        log=log)
+
+                    # FIXME: perhaps the dependency archive (the Cachi2) might be handled differently
+                    run(["tar", "xvf", extra_src_archive], check=True, cwd=dest_dir)
+                    os.unlink(f"{dest_dir}/{extra_src_archive}")
+
+                else:
+                    log.warning("No known operation happened on layer %s", digest)
+
+            return len(os.listdir(sib_dirs.rpm_dir)) > 0 or len(sib_dirs.extra_src_dirs) > 0
+
+
+        def gather_prefetched_sources(work_dir: str, cachi2_dir: str, sib_dirs: SourceImageBuildDirectories) -> bool:
+            log = logging.getLogger("source-build.prefetched-sources")
+
+            # Guess if hermetic build is enabled
+            # NOTE: this guess does depend on how cachi2 runs inside prefetch-dependencies task.
+            cachi2_output_dir = f"{cachi2_dir}/output"
+            cachi2_env = f"{cachi2_dir}/cachi2.env"
+
+            has_cachi2_output = os.path.exists(cachi2_output_dir) and os.path.isdir(cachi2_output_dir)
+            has_cachi2_env = os.path.exists(cachi2_env) and os.path.getsize(cachi2_env) > 0
+            if not (has_cachi2_output and has_cachi2_env):
+                log.info("Cannot discover artifacts generated by cachi2. Seems hermetic build is not enabled.")
+                return False
+
+            def _find_prefetch_source_archives():
```

Comment:
I don't understand this method at all

cachito output has directory structure

└── deps

    └── gomod

        └── pkg

            └── mod

                └── cache

                    └── download

                        ├── github.com
and in there directories for specific sources

Reply:
so what you should do is grab whole outputs/deps and tar.gz it and use that

Reply:
and I'd include also cachito2.env file there

Reply:
> and I'd include also cachito2.env file there

It is included. The code is below.

Reply:
Here is an example of the output directory:

```
cachi2-output2/
├── bom.json
└── deps
    ├── gomod
    │   └── pkg
    │       └── mod
    │           └── cache
    │               └── download
    │                   ├── github.com
    │                   │   ├── davecgh
    │                   │   │   └── go-spew
    │                   │   │       └── @v
    │                   │   │           ├── list
    │                   │   │           ├── v1.1.0.mod
    │                   │   │           └── v1.1.1.mod
    │                   │   ├── gogo
    │                   │   │   └── protobuf
    │                   │   │       └── @v
    │                   │   │           ├── list
    │                   │   │           ├── v1.3.2.info
    │                   │   │           ├── v1.3.2.lock
    │                   │   │           ├── v1.3.2.mod
    │                   │   │           ├── v1.3.2.zip
    │                   │   │           └── v1.3.2.ziphash
    ├── npm
    │   ├── accepts-1.3.8.tgz
    │   ├── array-flatten-1.1.1.tgz
    │   └── vary-1.1.2.tgz
    └── pip
        ├── Flask-2.3.1.tar.gz
        └── requests-2.30.0.tar.gz
```

`_find_prefetch_source_archives` gathers each of the archive and remain the path which is relative to the output directory. As a result, when extract sources from the source image, extracting all `extra-src-*.tar` will get the original output directory tree populated by cachi2.

> so what you should do is grab whole outputs/deps and tar.gz it and use that

Do you mean make a tar archive from `outputs/deps` and include all files under the `deps/` directory?

I remember we discussed that we want one source per layer. My understanding is one source means a source archive, a .tgz for npm, a .tar.gz for pip and a .zip for gomod.

Reply:
ok it makes sense now

so you are basically finding archives and putting them in the same path as in that cachito output

ok in that case it looks good now

Reply:
for Go, the other files should be metadata files. For the purpose of shipping sources, not sure how much helpful by including those metadata files.

The above example includes two different kinds of Go package directories, `go-spew` does not have archive inside, include it in the source image as well?

Reply:
Selecting files base on extension will be error-prone. For example, python sdists historically allowed not only `tar.gz`, but all of these extensions - all of which cachi2 supports. https://github.com/containerbuildsystem/cachi2/blob/84ac71f613e2f51becbc7d5b90ccf4c9c72e2768/cachi2/core/package_managers/pip.py#L56
```python
[".zip", ".tar.gz", ".tar.bz2", ".tar.xz", "tar.Z", ".tar"]
```
Once cachi2 implements Python wheel support, there will also be `*.whl` files (which are zip files but with a different extension). I don't remember what the decision was on those, IIRC we wanted to include them?

With Ruby support, there will be `*.gem` files (which are also zip files I think)

Reply:
And there was an idea floating around that source containers could contain everything needed for a re-build. In which case we should really dump the whole cachi2/ directory in the container, not filter anything out. @arewm did that idea progress anywhere?

Reply:
It is still an idea, but not one that has been written up into a feature and scoped. If we can achieve this or work towards it even, then great! If not, then we can defer until it becomes a priority. Additionally, if this work will be harder for some languages, it might need to be added with incremental language support.

Reply:
Hi, thank you all for feedback and information. This is the initial implementation for phase 1. I would keep it for now for the phase 1, then improve a little in the phase 2. I'll create a story for the archive type detention for the phase 2 that does not rely on the file extension as much as possible.

Reply:
After rethinking of this, I changed the way to detect the archive, that is based on mimetype. Then, when the requirements mentioned above are clear, we can reconsider a better way to dump the whole cachi2/ directory in the container. For the initial implementation, the task is able to cover most of the cases by not enumerating file extensions.